### ISSUE TITLE
3.x: 403 errors for moved asset files

### ISSUE THREAD

**Issue by xini** [silverstripe/silverstripe-framework#7690]
I am trying to implement 301 redirects for files stored in assets. I got into this issue when we rebuilt a site and discovered that we get a lot of 403 'not allowed' errors for files in the assets directory that have moved.
When a folder exists, but a file in that folder doesn't, you get a 404 error (e.g. `www.domain.tld/assets/folder/nofile.pdf`), as expected.
But when you hit an URL to a file in a non-existing folder, you get a 403 (e.g. `www.domain.tld/non-existing/no-file.pdf`).

The 403 errors are caused by ‘Deny from all’ in the assets/.htaccess file. This seems to have priority over any redirects added to the main .htaccess file for files that have changed path. It looks like the ‘Deny from all’ prohibits apache from checking whether the file actually exists.

To fix that, I have removed the global ‘Deny from all’ and changed the regex to the following using a negation:
```
<FilesMatch ".(!(?i:html|htm|xhtml|js|css|bmp|png|gif|jpg|jpeg|ico|pcx|...|jar|xml|pdf|gpx|kml))$">
Deny from all
</FilesMatch>
```
Is that safe to use? Should that be changed in core?

Also, any 404 handlers in SS (e.g. misdirection, etc) don't work for asset files because of 
```
RewriteCond %{REQUEST_FILENAME} !-f
```
in the root .htaccess file. 

If I add
```
RewriteCond %{REQUEST_URI} ^assets/ [NC,OR]
```
right before that line in the main .htaccess file, this works, but of course it also means that every request to a file goes through the whole SS controller stuff, which makes the site slower. With this the 404 handlers also works for asset files.

**Comment by tractorcow** [comment]
> Is that safe to use? Should that be changed in core?

We probably aren't going to do any major changes to asset routing in 3.x; Given that 4.x has an entirely new asset system, we're likely to focus our efforts on optimising this.

Users are advised to, as you have done, manually modify and optimise your sites' root and assets `.htaccess` file in a way appropriate for your project.

If you can find a way to resolve the directory name issue causing 403 instead of 404 then we could PR it to the installer, but the fix you've tested seems to turn all disallowed file extensions into 404s instead of 403s (e.g. /assets/not-a-folder/program.exe -> 404).

**Comment by xini** [comment]
Hi @tractorcow. Thanks. My fix is only turning requests to non existing files to 404 (e.g. your example of /assets/not-a-folder/program.exe), which is correct. If the file exists and the extension isn't allowed, you get a 403 (e.g. /assets/existing-folder/existing-file.exe).

**Comment by tractorcow** [comment]
Hm, I'd gotten a 404 not a 403. Perhaps I had not replicated your `.htaccess` correctly? Can you put it into a gist verbatim for me and I'll test it.

**Comment by tractorcow** [comment]
I'll just temporarily open this... not confirming this is a bug yet just want to finish investigation. :)

**Comment by xini** [comment]
Thanks. Here the gist: https://gist.github.com/xini/502d988e7ccf1140d9bebd23be4be169

**Comment by tractorcow** [comment]
Oh, with this `.htaccess` I'm able to download `.exe` files without a 403.
Non-existing `.exe` queries are reported as 404.
Back to the drawing board?

**Comment by tractorcow** [comment]
I'm on apache 2.4 by the way.

**Comment by dhensby** [comment]
Is there a problem with returning 404 for forbidden extensions? Isn't it a type of disclosure vulnerability otherwise?

**Comment by tractorcow** [comment]
It doesn't seem to be consistent either way... if the path itself is forbidden, it's not a disclosure. If the specific file is forbidden, yes it is a disclosure. Thus all forbidden extensions should be 403 regardless of whether they exist.

**Comment by tractorcow** [comment]
I'm marking this as a bug, for now, but I'm not 100% certain what the state of this bug is. It feels low priority, so we can come back to this and investigate it later.

As it stands, there is inconsistent behaviour in reporting of 403 / 404 for missing / disallowed assets in 3.x.

**Comment by xini** [comment]
@tractorcow you're right, my .htaccess didn't work. sorry. 

I think showing a 404 for everything except the allowed files would be preferable. 

I have been digging into this, but I can't find a solution to not showing a 403 if a file doesn't exist.

I have come up with an update to the regex using a negative lookbehind (based on https://stackoverflow.com/questions/12926604/inverted-filesmatch):
```
Allow from all
<FilesMatch "(?<!i:\.html|\.htm|\.xhtml|\.js|\.css|\.bmp|\.png|\.gif|\.jpg|\.jpeg|\.ico|\.pcx|\.tif|\.tiff|\.au|\.mid|\.midi|\.mpa|\.mp3|\.ogg|\.m4a|\.ra|\.wma|\.wav|\.cda|\.avi|\.mpg|\.mpeg|\.asf|\.wmv|\.m4v|\.mov|\.mkv|\.mp4|\.ogv|\.webm|\.swf|\.flv|\.ram|\.rm|\.doc|\.docx|\.dotx|\.dotm|\.txt|\.rtf|\.xls|\.xlsx|\.xltx|\.xltm|\.pages|\.ppt|\.pptx|\.potx|\.potm|\.pps|\.csv|\.cab|\.arj|\.tar|\.zip|\.zipx|\.sit|\.sitx|\.gz|\.tgz|\.bz2|\.ace|\.arc|\.pkg|\.dmg|\.hqx|\.jar|\.xml|\.pdf|\.gpx|\.kml)$">
	Deny from all
</FilesMatch>
```
But it still shows a 403 for all files except the existing allowed ones. Where is my mistake in this one? This should work, shouldn't it?

**Comment by tractorcow** [comment]
If you don't want to show 403 then you can't use `deny`, pretty much.

It's ok to show 403, as long as it skips the file-exists check.

**Comment by xini** [comment]
yeah, but how do I do that? I need redirects in the main htaccess to work for non-existing files, but the redirects don't work because of the deny.

**Comment by tractorcow** [comment]
For your app, remove the filesmatch, and use exclusively rewriterules to serve either 200, redirect, or manual 403 for disallowed extensions.

**Comment by tractorcow** [comment]
```
RewriteRule "\.exe" "-" [F]
```

See http://httpd.apache.org/docs/current/rewrite/flags.html#flag_f

**Comment by xini** [comment]
Here a different solution for the '/assets.htaccess' file:

```
ErrorDocument 404 /error-404.html
ErrorDocument 500 /error-500.html

Allow from all
<FilesMatch "(?<!i:\.html|\.htm|\.xhtml|\.js|\.css|\.bmp|\.png|\.gif|\.jpg|\.jpeg|\.ico|\.pcx|\.tif|\.tiff|\.au|\.mid|\.midi|\.mpa|\.mp3|\.ogg|\.m4a|\.ra|\.wma|\.wav|\.cda|\.avi|\.mpg|\.mpeg|\.asf|\.wmv|\.m4v|\.mov|\.mkv|\.mp4|\.ogv|\.webm|\.swf|\.flv|\.ram|\.rm|\.doc|\.docx|\.dotx|\.dotm|\.txt|\.rtf|\.xls|\.xlsx|\.xltx|\.xltm|\.pages|\.ppt|\.pptx|\.potx|\.potm|\.pps|\.csv|\.cab|\.arj|\.tar|\.zip|\.zipx|\.sit|\.sitx|\.gz|\.tgz|\.bz2|\.ace|\.arc|\.pkg|\.dmg|\.hqx|\.jar|\.xml|\.pdf|\.gpx|\.kml)$">
	RewriteRule .* - [R=404,NC,L]
</FilesMatch>
```

It still uses a white list, it shows a 404 for everything except existing allowed files, and 404 handlers in the CMS like misdirection work too. This would prevent any information leak for existing but disallowed files.

Unfortunately I had to re-add the error documents, otherwise it shows the ugly default 404 page from Apache. Does anyone see a different solution for this?
